### PR TITLE
Document how addons can listen for file changes

### PR DESCRIPTION
**Pull request by Earlopain**
### Motivation

Addons may want to listen for changes to files and do something about it. `ruby-lsp-rails` for example listens for `structure.sql` to restart the rails runner for updated column information, after https://github.com/Shopify/ruby-lsp-rails/pull/341 is merged.

`ruby-lsp` itself wants to check for changes to the rubocop config in the server to resolve https://github.com/Shopify/ruby-lsp/issues/1457

`rubyfmt` for `ruby-lsp-rubyfmt` appears to change behaviour depending on `.gitignore` and `.rubyfmtignore`.

----

I noticed that some activation methods didn't yet take the global state in the docs here . I've updated these in a separate commit here.

### CODE REVIEW

File: ADDONS.md [hunk 1]
```diff
@@ -301,6 +301,66 @@ module RubyLsp
 end
 ```
 
+### Registering for file update events
+
+By default, the Ruby LSP listens for changes to files ending in `.rb` to continuously update its index when Ruby source
+code is modified. If your your addon uses a tool that is configured through a file (like RuboCop and its `.rubocop.yml`)
```
Comment (andyw8) [inline ADDONS.md:307]:
```suggestion
code is modified. If your addon uses a tool that is configured through a file (like RuboCop and its `.rubocop.yml`)
```

File: ADDONS.md [hunk 2]
```diff
@@ -301,6 +301,66 @@ module RubyLsp
...
+  module MyGem
+    class Addon < ::RubyLsp::Addon
+      def activate(global_state, message_queue)
+        register_additional_file_watchers(global_state, message_queue)
+      end
+
+      def deactivate; end
+
+      def register_additional_file_watchers(global_state, message_queue)
+      # Clients are not required to implement this capability
```
Comment (andyw8) [inline ADDONS.md:324]: Let's fix the indentation here.

File: ADDONS.md [hunk 3]
```diff
@@ -301,6 +301,66 @@ module RubyLsp
...
+      return unless global_state.supports_watching_files
+
+        message_queue << Request.new(
```
Comment (andyw8) [inline ADDONS.md:327]: This is good to document but illustrates that Ruby LSP should probably provide a better API for this.

File: ADDONS.md [hunk 4]
```diff
@@ -301,6 +301,66 @@ module RubyLsp
...
+you can register for changes to these files and react when the configuration changes.
+
+**Note**: you will receive events from `ruby-lsp` and other addons as well, in addition to your own registered ones.
+
+
+```ruby
+module RubyLsp
+  module MyGem
+    class Addon < ::RubyLsp::Addon
+      def activate(global_state, message_queue)
```
Comment (andyw8) [inline ADDONS.md:317]: In the other code samples we include the Sorbet signatures, so I'm wondering if we should do the same here for consistency.

**Review by andyw8 — APPROVED** [review]
Looks good to me but I'll like @vinistock to check it over too.

File: ADDONS.md [hunk 5]
```diff
@@ -301,6 +301,66 @@ module RubyLsp
...
+      end
+
+
+      def workspace_did_change_watched_files(changes)
+        if changes.any? { |change| change[:uri].end_with?(".my-config.yml") }
+          # Do something to reload the config here
+        end
+      end
+
+      def name
```
[the remark this answers is not in the document]
Comment (Earlopain) [reply ADDONS.md:356]: All the other examples have it as well. But I can drop it here, sure. It doesn't add much

File: ADDONS.md (hunk 2 above)
Comment (Earlopain) [reply ADDONS.md:324]: Good catch. Would you be open to add [rubocop-md](https://github.com/rubocop/rubocop-md) in a separate PR?

Comment (andyw8) [reply ADDONS.md:324]: Sure, I haven't tried it before but it looks useful.

File: ADDONS.md (hunk 3 above)
Comment (vinistock) [reply ADDONS.md:327]: I agree that the API is not the friendliest, but let's move forward with the way it is now and see if we start seeing more use cases of addons that need to watch files.

**Review by vinistock — APPROVED** [review]

**Comment by vinistock** [comment]
Thank you for the contribution!

**Pull request merged**